File: src/client/java/com/example/client/config/ConfigCategory.java
package com.example.client.config;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BooleanSupplier;
import java.util.function.Consumer;

public class ConfigCategory {

    //   FIELDS (Category name and its contained list of toggle options)
    public final String name;
    public final List<ToggleEntry> entries = new ArrayList<>();

    //   CONSTRUCTOR
    public ConfigCategory(String name) {
        this.name = name;
    }

    //   BUILDER METHODS (Used to chain settings onto a category easily)
    public ConfigCategory toggle(String label, BooleanSupplier getter, Consumer<Boolean> setter) {
        entries.add(new ToggleEntry(label, getter, setter));
        return this;
    }

    //   SUBCLASS: TOGGLE ENTRY (Data holder for an individual config option)
    public static class ToggleEntry {
        public final String label;
        public final BooleanSupplier getter;
        public final Consumer<Boolean> setter;

        public ToggleEntry(String label, BooleanSupplier getter, Consumer<Boolean> setter) {
            this.label = label;
            this.getter = getter;
            this.setter = setter;
        }
    }
}

File: src/main/kotlin/com/cbza/net/config/ConfigCategory.kt
package com.cbza.net.config

import kotlin.reflect.KMutableProperty0

class ConfigCategory(val name: String) {
	val entries = mutableListOf<ToggleEntry>()

	fun toggle(
		label: String,
		property: KMutableProperty0<Boolean>,
		description: String = ""
	): ConfigCategory {
		entries.add(ToggleEntry(label, { property.get() }, { property.set(it) }, description))
		return this
	}

	fun toggle(
		label: String,
		getter: () -> Boolean,
		setter: (Boolean) -> Unit,
		description: String = ""
	): ConfigCategory {
		entries.add(ToggleEntry(label, getter, setter, description))
		return this
	}

	class ToggleEntry(
		val label: String,
		val getter: () -> Boolean,
		val setter: (Boolean) -> Unit,
		val description: String = ""
	)
}